include/h2o/httpclient_internal_h2.h

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2017 Ichito Nagata, Fastly, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+#ifndef h2o__http2client_h
+#define h2o__http2client_h
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "khash.h"
+#include "h2o/httpclient_internal.h"
+#include "h2o/http2_common.h"
+
+enum enum_h2o_http2client_stream_state {
+    H2O_HTTP2CLIENT_STREAM_STATE_SEND_HEADERS,
+    H2O_HTTP2CLIENT_STREAM_STATE_SEND_BODY,
+    H2O_HTTP2CLIENT_STREAM_STATE_RECV_HEADERS,
+    H2O_HTTP2CLIENT_STREAM_STATE_RECV_BODY,
+};
+
+enum enum_h2o_http2client_conn_state {
+    H2O_HTTP2CLIENT_CONN_STATE_OPEN,
+    H2O_HTTP2CLIENT_CONN_STATE_HALF_CLOSED,
+    H2O_HTTP2CLIENT_CONN_STATE_IS_CLOSING,
+};
+
+struct st_h2o_http2client_stream_t;
+KHASH_MAP_INIT_INT64(stream, struct st_h2o_http2client_stream_t *)
+
+struct st_h2o_http2client_conn_t {
+    h2o_httpclient_ctx_t *ctx;
+    h2o_url_t origin_url;
+    h2o_socket_t *sock;
+    enum enum_h2o_http2client_conn_state state;
+    khash_t(stream) * streams;
+    h2o_linklist_t link;
+    h2o_http2_settings_t peer_settings;
+    uint32_t max_open_stream_id;
+    size_t num_streams;
+    h2o_timer_t io_timeout;
+    h2o_timer_t keepalive_timeout;
```

Comment:
Can't we use one timer object instead of having two?

My understanding is that their use is exclusive; i.e. keepalive_timeout is unset when io_timeout is used, io_timeout is unset when keepalive_timeout is used.